### PR TITLE
DROOLS-4395: [DMN Designer] Validation fails for included PMML model

### PR DESCRIPTION
**Pull request by karreiro**
See https://issues.jboss.org/browse/DROOLS-4395

### CODE REVIEW

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Using the following PMML file gave the following error in Business Central:
```
15:25:06,400 ERROR [org.kie.workbench.common.services.backend.builder.core.Builder] (EJB default - 2) null: java.lang.NullPointerException
	at org.drools.compiler.kie.builder.impl.KieBuilderSetImpl.findResourcesWithMessages(KieBuilderSetImpl.java:127)
	at org.drools.compiler.kie.builder.impl.KieBuilderSetImpl.registerInitialErrors(KieBuilderSetImpl.java:78)
	at org.drools.compiler.kie.builder.impl.KieBuilderSetImpl.<init>(KieBuilderSetImpl.java:68)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.createFileSet(KieBuilderImpl.java:788)
	at org.kie.workbench.common.services.backend.builder.core.Builder.buildIncrementally(Builder.java:533)
	at org.kie.workbench.common.services.backend.builder.core.Builder.addResource(Builder.java:354)
	at org.kie.workbench.common.services.backend.builder.core.Builder.addResource(Builder.java:320)
	at org.kie.workbench.common.services.backend.builder.core.Builder.updateResource(Builder.java:414)
	at org.kie.workbench.common.services.backend.builder.core.BuildHelper.updatePackageResource(BuildHelper.java:167)
	at org.kie.workbench.common.services.backend.builder.core.BuildHelper$Proxy$_$$_WeldClientProxy.updatePackageResource(Unknown Source)
	at org.kie.workbench.common.services.backend.builder.ala.LocalBuildExecConfigExecutor.apply(LocalBuildExecConfigExecutor.java:70)
	at org.kie.workbench.common.services.backend.builder.ala.LocalBuildExecConfigExecutor.apply(LocalBuildExecConfigExecutor.java:32)
	at org.kie.workbench.common.services.backend.builder.ala.LocalBuildExecConfigExecutor$Proxy$_$$_WeldClientProxy.apply(Unknown Source)
	at org.guvnor.ala.pipeline.execution.PipelineExecutor.lambda$continuePipeline$0(PipelineExecutor.java:109)
	at org.guvnor.ala.pipeline.StageUtil$1.execute(StageUtil.java:38)
	at org.guvnor.ala.pipeline.StageUtil$1.execute(StageUtil.java:33)
	at org.guvnor.ala.pipeline.execution.PipelineExecutor.continuePipeline(PipelineExecutor.java:94)
	at org.guvnor.ala.pipeline.execution.PipelineExecutor.execute(PipelineExecutor.java:76)
	at org.kie.workbench.common.services.backend.builder.ala.BuildPipelineInvoker.invokeLocalBuildPipeLine(BuildPipelineInvoker.java:88)
	at org.kie.workbench.common.services.backend.builder.ala.BuildPipelineInvoker$Proxy$_$$_WeldClientProxy.invokeLocalBuildPipeLine(Unknown Source)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceHelper.invokeLocalBuildPipeLine(BuildServiceHelper.java:163)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceHelper.localBuild(BuildServiceHelper.java:98)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceHelper$Proxy$_$$_WeldClientProxy.localBuild(Unknown Source)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceImpl.buildIncrementally(BuildServiceImpl.java:136)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceImpl.updatePackageResource(BuildServiceImpl.java:126)
	at org.kie.workbench.common.services.backend.builder.service.BuildServiceImpl$Proxy$_$$_WeldClientProxy.updatePackageResource(Unknown Source)
	at org.guvnor.common.services.builder.ResourceChangeIncrementalBuilder$4.execute(ResourceChangeIncrementalBuilder.java:262)
	at org.guvnor.common.services.builder.IncrementalBuilderExecutorManager.execute(IncrementalBuilderExecutorManager.java:90)
```
I doubt it's related to this PR, but just mentioning.

@tarilabs Hi, please see above. Do you know if there are any specific conditions we need to handle PMML files in Business Central? Should `KieBuilder` handle their compilation "out of the box" or do I need to ensure _jpmml_ is included!?

@danielzhe Perhaps when you review you could also check this.

[test_scorecard.txt](https://github.com/kiegroup/kie-wb-common/files/3566398/test_scorecard.txt)

You'll need to rename `.pmml` as I had to rename to `.txt` to upload here.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelper.java [hunk 1]
```diff
@@ -84,4 +87,22 @@
...
                                                                final String namespace);
+
+    /**
+     * This method finds the {@link Path} of DMN model.
+     * @param metadata represents the metadata from the a DMN model from the scanned project.
+     * @param modelNamespace represents the namespace of the desired DMN model.
+     * @param modelName represents the mode name of the desired DMN model.
+     * @return the {@link Path} of the desired DMN model.
+     */
+    Path getModelPath(final Metadata metadata,
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelper.java]: This only returns the `Path` of **DMN** file not a PMML import.

**Comment by tarilabs** [comment]
> Using the following PMML file gave the following error in Business Central

If that is happening with say, and empty project and just uploading the .pmml file, it would be best to check with @lanceleverich ?
I prefer to mention as just by looking at the stacktrace I see no dmn package of the integration code.
AFAIK if the .pmml file is a supported one, it should just compile ootb

**Comment by karreiro** [comment]
Thank you @manstis.

I've fixed the `getModelPath` (I've renamed it to `getDMNModelPath`).

For a more deep refactoring, I've created this JIRA https://issues.jboss.org/browse/DROOLS-4499 I do agree with the points you've raised.

Regarding this [error](https://github.com/kiegroup/kie-wb-common/pull/2848#issuecomment-527168279) I think we can check asynchronously with @tarilabs and @lanceleverich (to not hold this PR). Wdyt?

Thanks again for the review :-)

**Comment by manstis** [comment]
@tarilabs @lanceleverich Sorry, I am forgetful.

Does a `pmml` file extension lead to compilation by `kie-pmml` or `jpmml`? For the recent demo I thought PMML handling moved to `jpmml` as `kie-pmml` did not support all the model types (or something like that). Was this just temporary for the demo? The DMN editor still uses `org.kie.dmn.core.pmml.PMMLInfo` that uses `jpmml`.

**Comment by danielezonca** [comment]
@manstis 
The answer is both :smile: 
We have a routine that checks if jpmml integration class is available in the classpath and if so it uses it, otherwise kie-pmml. And you are right, DMN Editor will import all models even if kie-pmml impl is not supporting it. I think this is fine for now, maybe we can raise a warning for the user in the future

**Comment by tarilabs** [comment]
@manstis Daniele's answer is correct, but I believe I should also clarify couple of points.

> For the recent demo I thought PMML handling moved to jpmml as kie-pmml did not support all the model types (or something like that). Was this just temporary for the demo? 

Yes that hard override was done only for the purpose of the demo.
On master we support both runtimes with the mechanism as Daniele explained and also w/ Vlad implemented.
Upcoming doc update will clarify also this.

> The DMN editor still uses org.kie.dmn.core.pmml.PMMLInfo that uses jpmml.

There might be some confusion here. We are using a "marshaller" component of jpmml for that scope, BSD-CL3 licensed, and for that scope only. However only that part won't provide any execution/evaluations, which requires the other jpmml integration modules. That is used only to "inspect" the content of the pmml file in a common way, which will later be need proper evaluation runtime support, by default with kie-pmml.

Hope this clarifies.

**Pull request merged**